patches/tModLoader/Terraria/ModLoader/PosData.cs

```diff
@@ -0,0 +1,166 @@
+using Microsoft.Xna.Framework;
+using System;
+using System.Collections.Generic;
+
+namespace Terraria.ModLoader
+{
+	public readonly struct PosData<T>
+	{
+
+		// Enumeration class to build an Ordered Sparse Lookup system
+		public class OrderedSparseLookupBuilder 
+		{
+			private readonly List<PosData<T>> list;
+			private PosData<T> last;
+
+			public OrderedSparseLookupBuilder(int capacity = 1048576) {
+				list = new List<PosData<T>>(capacity);
+				last = new PosData<T>(-1, default);
+			}
+
+			public void Add(int x, int y, T value) => Add(PosData.CoordsToPos(x, y), value);
+
+			public void Add(int pos, T value) {
+				if (pos <= last.pos)
+					throw new ArgumentException($"Must build in ascending index order. Prev: {last.pos}, pos: {pos}");
+
+				if (!EqualityComparer<T>.Default.Equals(value, last.value))
+					list.Add(last = new PosData<T>(pos, value));
+			}
+
+			public PosData<T>[] Build() => list.ToArray();
+		}
+
+		// Enumeration class to access data in a Sparse Lookup System
+		public class OrderedSparseLookupReader
+		{
+			private readonly PosData<T>[] data;
+			private PosData<T> current;
+			private int nextIdx;
+
+			public OrderedSparseLookupReader(PosData<T>[] data) {
+				this.data = data;
+				current = new PosData<T>(-1, default);
+				nextIdx = 0;
+			}
+
+			public T Get(int x, int y) => Get(PosData.CoordsToPos(x, y));
+
+			public T Get(int pos) {
+				if (pos <= current.pos)
+					throw new ArgumentException($"Must read in ascending index order. Prev: {current.pos}, pos: {pos}");
+
+				while (nextIdx < data.Length && data[nextIdx].pos <= pos)
+					current = data[nextIdx++];
+
+				return current.value;
+			}
+		}
+
+		public readonly int pos;
+		public readonly T value;
+
+		public int X => pos / Main.maxTilesY;
+		public int Y => pos % Main.maxTilesY;
+
+		public PosData(int pos, T value) {
+			this.pos = pos;
+			this.value = value;
+		}
+
+		public PosData(int x, int y, T value) : this(PosData.CoordsToPos(x, y), value) { }
+	}
+
+	public static class PosData
+	{
+		/// <summary>
+		/// Gets a Position ID based on the x,y position. If using in an order sensitive case, see NextLocation.
+		/// </summary>
+		/// <param name="posX"></param>
+		/// <param name="posY"></param>
+		/// <returns></returns>
+		public static int CoordsToPos(int x, int y) => x * Main.maxTilesY + y;
+
+
+		/// <summary>
+		/// Searches for the interval posMap[i].posID < provided posID < posMap[i + 1].posID.
+		/// </summary>
+		/// <param name="posMap"></param>
+		/// <param name="posID"></param>
+		/// <returns></returns>
+		public static int FindIndex<T>(this PosData<T>[] posMap, int pos) {
+			if (posMap.Length == 0) {
+				throw new ArgumentException($"Can't find the index in an empty posMap. Please verify map is non-empty before calling.");
+			}
```

Comment:
please remove this condition. I updated the iteration to return -1 when `pos` is < the first entry in the map (or the map is empty)

This works naturally with the builder, as there's essentially a phantom element at `-1` with the value of `default(T)` this means you don't ever have to check that the map is non-empty.

Conceptually, an empty map means that the entire world is filled with `default(T)`, which is a sensible thing to work with